### PR TITLE
[BC-169] Advertise actual IP address

### PR DESCRIPTION
**Pull request by cemozerr**
## Fixed Issue(s)
https://pegasys1.atlassian.net/browse/BC-169

### CODE REVIEW

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java [hunk 1]
```diff
@@ -153,6 +160,19 @@ public LibP2PNetwork(
             });
   }
 
+  public Multiaddr composeAdvertisedAddr(NetworkConfig config) throws UnknownHostException {
+    String ip;
+    if (!config.getAdvertisedIp().trim().isEmpty()) {
+      ip = config.getAdvertisedIp();
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: We probably also want to check that this at least looks like a valid ip: `InetAddresses.isInetAddress(config.getAdvertisedIp())`

**Review by mbaxter — APPROVED** [review]

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java [hunk 2]
```diff
@@ -80,7 +83,12 @@ public LibP2PNetwork(
     this.privKey = config.getPrivateKey();
     this.nodeId = new LibP2PNodeId(PeerId.fromPubKey(privKey.publicKey()));
 
-    advertisedAddr = new Multiaddr("/ip4/127.0.0.1/tcp/" + config.getAdvertisedPort());
+    try {
+      advertisedAddr = composeAdvertisedAddr(config);
+    } catch (UnknownHostException err) {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: I'd probably move the exception handling to the `composeAdvertisedAddr` method

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java [hunk 3]
```diff
@@ -153,6 +161,19 @@ public LibP2PNetwork(
             });
   }
 
+  public Multiaddr composeAdvertisedAddr(NetworkConfig config) throws UnknownHostException {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: Maybe just:
```suggestion
  public Multiaddr getAdvertisedIp(NetworkConfig config) throws UnknownHostException {
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/NetworkConfig.java [hunk 4]
```diff
@@ -45,6 +50,15 @@ public NetworkConfig(
       final boolean logMuxFrames) {
     this.privateKey = privateKey;
     this.networkInterface = networkInterface;
+
+    if (advertisedIp.trim().isEmpty()) {
+      this.advertisedIp = Optional.empty();
+    } else if (!isInetAddress(advertisedIp)) {
+      throw new RuntimeException("Advertised ip is set incorrectly.");
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/NetworkConfig.java]:
```suggestion
      throw new IllegalArgumentException("Advertised ip is set incorrectly.");
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java [hunk 5]
```diff
@@ -153,6 +161,19 @@ public LibP2PNetwork(
...
+    String ip;
+    if (config.getAdvertisedIp().isPresent()) {
+      ip = config.getAdvertisedIp().get();
+    } else if (NetworkUtility.isNetworkInterfaceAvailable(config.getNetworkInterface())) {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]:
```suggestion
    } else if (!NetworkUtility.isUnspecifiedAddress(config.getNetworkInterface())) {
```

File: util/src/main/java/tech/pegasys/artemis/util/network/NetworkUtility.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.Enumeration;
+import java.util.function.Supplier;
+
+public class NetworkUtility {
+  public static final String INADDR_ANY = "0.0.0.0";
```
Comment (mbaxter) [inline util/src/main/java/tech/pegasys/artemis/util/network/NetworkUtility.java:17]: (optional) Might be nice to replace "0.0.0.0" strings in the codebase with this constant

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java (hunk 5 above)
Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: done.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/NetworkConfig.java (hunk 4 above)
Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/NetworkConfig.java]: done.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java (hunk 3 above)
Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: done.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java (hunk 2 above)
Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PNetwork.java]: done.

File: util/src/main/java/tech/pegasys/artemis/util/network/NetworkUtility.java (hunk 6 above)
Comment (cemozerr) [reply util/src/main/java/tech/pegasys/artemis/util/network/NetworkUtility.java:17]: done.

**Pull request merged**